Makefile

```diff
@@ -1,18 +1,23 @@
-.PHONY: all
-all: bin/deployer containers
+.DEFAULT_GOAL := help
 
-bin/deployer:
-	$(MAKE) -C deployer
+.PHONY: all
+all: bin/deployer containers ## build everything
 
-.PHONY: containers
-containers:
-	$(MAKE) -C containers
+bin/deployer: deployer/deployer.bin ## build deployer tool from source
+	mkdir -p bin
+	cp deployer/deployer.bin bin/deployer
+deployer/deployer.bin:
+	$(MAKE) -C deployer deployer.bin
```

Comment:
I rather like these changes. I wouldn't have thought of them. Thanks @jpluscplusm! :-)

I'm curious about `deployer.bin` as a place to build to. It's good to make `deployer/Makefile` independent of its parent directory, but building to `deployer.bin` seems kind of ugly. I guess though that a `build` directory would be unnecessary, and a `.bin` extension means it can be `.gitignore`d. Is there a history to this/thoughts you could share so I can learn? Thanks!

Reply:
@46bit Dagnamit - I totally forgot to put this on my list of things to reply to!

The reasoning behind the `.bin` suffix is this: you already had a `deployer` entry in the `deployer` directory. My local golang setup is sufficiently borked that I can't do a build to learn which binary artefact gets created if the `go build` step doesn't have a `-o` param. I /guessed/ it might be `deployer`, which would clash, so opted for the `.bin` suffix instead.

`..` in a Makefile always smells fishy to me, and removing that was of course the main aim :-)